### PR TITLE
[READY] Space Suits use cells and warm the wearer

### PR DESCRIPTION
**Pull request by nightred**
## About The Pull Request

This makes space suits use a cell and warm the wearer, This was handled by the handle_environment proc in the past with an override to make the suits work. This corrects this issue, and removes the override.

All space suits will have a cell installed at round start, this cell will last 15 min on average usage.

The cell cover can be toggled by alt+clicking (press the button), or using a crowbar (just pry it open) on the suit.
With the cover open you can remove the cell by using the suit in hand, or ctrl+clicking the suit. 
As well with the cover open you can use a screwdriver to set the thermal regulator's temperature. The normal range is 20 c to 45 c, do not use an emag on the suit to increase the temperature range.

Examining the suit will report the temperature setting and the cells charge, if the cover is open you will also be able to see the type of cell that is in the suit.

Suit storage units will now charge the cells in suits, they will also require power to operate properly now.

Here is a clip of some basic suit action.
[Space Suits](https://youtu.be/F6DyHFmCK4g)

Still to come:
- [x] toggle on/off
- [x] action buttons
- [x] suit storage charges suits
- [x] emp shielding
- [x] emag act
- [x] hud status
- [x] issues with some suit action buttons
- [x] balance emp protection
- [x] add shielded cells in engineering on all stations
- [x] final testing

## Why It's Good For The Game

* Space suits should be managing the wearers temp in space not using some override.
* Lizards now have a portable heater in the form of any space suit.
* Now that suits need a cell it is not as easy to spend all session in a space suit.
* New interactions with hardsuits and emp's give nice combat buff to an underused item.
* Nuke ops can raid xeno for some nice yellow cells, or science for a big old bluespace cell.

## Changelog
:cl:
add: Space Suits warm the wearer, and use Cells
add: Suit Storage charges Space Suits
add: EMP's cause Hard Suit's to burn the wearer
add: Engineering now has emp proof cells with the suits
add: Hud status for space suit power level
add: ERT hardsuits get EMP protection AND ONLY THEM
/:cl:

### CODE REVIEW

**Comment by SkeletalElite** [comment]
Do they function like they do currently if the cell runs out or is removed?

**Comment by nightred** [comment]
> Do they function like they do currently if the cell runs out or is removed?

They do not, the override that keeps your temperature stable will be removed.
You will still get the insulation of the suit but that will not protect you from space cold, or stop you from overheating on the station for long.

**Comment by AdamElTablawy** [comment]
Might as well raise the temp setting limit some, so that people could sabotage a space suit by setting its internal heating to extremely high/extremely low temperatures.

**Comment by Ohagi-Chan** [comment]
Can you wear a suit with the cell turned off, or do you need to open the panel, remove the suit, place a cell inside and put the suit back on?

**Comment by itseasytosee** [comment]
II would be fine with this if there was a hud toggle for the heating. constant heating while having the insulation of a spacesuit seems wasteful.

**Comment by nightred** [comment]
>II would be fine with this if there was a hud toggle for the heating.

There is a hub button for heating on/off. Sprites incoming.

File: code/modules/clothing/spacesuits/chronosuit.dm [hunk 1]
```diff
@@ -178,8 +179,6 @@
 						camera.remove_target_ui()
 			else
 				new_camera(user)
-	else
```
Comment (ATH1909) [inline code/modules/clothing/spacesuits/chronosuit.dm]: this won't break chronosuits, right?

Comment (nightred) [reply code/modules/clothing/spacesuits/chronosuit.dm]: This was done so Chrono suits will not break.

**Comment by Tlaltecuhtli** [comment]
cells inside things suffer from being deleted if the suit gets any ex_act(), is there any code handling this situation?  #48020

**Comment by AWalton** [comment]
Security, Syndicate and ERT hardsuits should get EMP shielding so the power cell isn't fried, as they're likely to come into contact with EMPs in combat. 

Also Syndicate and ERT hardsuits need yellow slime cells since you're *supposed* to be wearing these suits for the entire duration of your mission.

**Comment by zxaber** [comment]
>action button

Why must you clutter the top of my screen? Is it possible that this could just be a crowbar to pop the panel open?

**Comment by nightred** [comment]
> > action button
> 
> Why must you clutter the top of my screen? Is it possible that this could just be a crowbar to pop the panel open?

Space suits use many actions already, jet packs use screwdrivers etc. This was done for speed of activation for the user, and most space suits already had several buttons and the style was kept.

**Comment by zxaber** [comment]
The action button on hardsuits is for the helmet, which can be expected to toggle semi-often for the functions of eating food, or getting a heal injection, or even to prove you aren't holding someone else's ID. For heated suits, I assume the battery doesn't need to be changed super regularly, and it seems like you have to remove the suit to actually swap the battery anyway. So this button will be both a rarely-used icon cluttering the action button area, and also not a quick-swap function like enabling a helmet or turning on the PDA light.

I'm not really a huge fan of action buttons in general, but this really seems like a task for an in-game tool (crowbar).

To be clear, I'm only talking about the pop-hatch button. The toggle action button for turning on/off the heating function makes sense.

**Comment by PKPenguin321** [comment]
You could probably put cell cover toggling on Alt click instead of an action button

**Comment by AWalton** [comment]
Thanks for the speedy changes. Also Captain's (Mark II) and SWAT (Mark I) hardsuits should probably have EMP shielding as well as they're also intended to see combat situations.

**Comment by nightred** [comment]
>To be clear, I'm only talking about the pop-hatch button. The toggle action button for turning on/off the heating function makes sense.

> You could probably put cell cover toggling on Alt click instead of an action button

Alt click to open, once more to remove the cell, and a third time to close the cover.
Now to you can remove the cell without taking it off the suit, and the action button is gone.

> Thanks for the speedy changes. Also Captain's (Mark II) and SWAT (Mark I) hardsuits should probably have EMP shielding as well as they're also intended to see combat situations.

Fixed up more EMP stuff.

**Comment by Ohagi-Chan** [comment]
Remove cell and close hatch are both on the alt click?
How do you close the hatch with a cell inside?

**Comment by CRITAWAKETS** [comment]
Engineering hardsuits also need EMP protection as the singulo shits out EMPs.

**Comment by Jalleo** [comment]
Singulo. Anomolies. Tesla. This will not protect the headsets on the user will it? (Dont let it protect headsets please

**Comment by nightred** [comment]
> Engineering hardsuits also need EMP protection as the singulo shits out EMPs.

This will be added, but to many are starting to get EMP protection.

> Remove cell and close hatch are both on the alt click?

Yes this sounded good when i did it, I am moving removal of the cell to ctrl. All on alt sounded so good in my head.

**Comment by zxaber** [comment]
Removal of an unnecessary action button is already a victory in my book, but why the avoidance of just having a crowbar for the hatch? Similar is already used for hatches on borgs and mechs, and it would avoid having to use two different modifier clicks.

**Comment by nightred** [comment]
> Removal of an unnecessary action button is already a victory in my book, but why the avoidance of just having a crowbar for the hatch? Similar is already used for hatches on borgs and mechs, and it would avoid having to use two different modifier clicks.

I can add crowbar also, but the issue is that the eva suits, security suits, medical, etc do not need a crowbar to do their job and forcing everyone to carry crowbar is a bad idea. These should be accessible and usable without any fancy tools in basic operating situations.

**Comment by trollbreeder** [comment]
Has this been tested with species that have funky nonstandard temperatures?

**Comment by nightred** [comment]
> Has this been tested with species that have funky nonstandard temperatures?

This has been made to work with all race start races and the body temps they have. Also a simple screwdriver will let you change the temperature of the suit at will.

**Comment by Fikou** [comment]
the 2 wizard hardsuits dont get emp protection

**Comment by nightred** [comment]
> the 2 wizard hardsuits dont get emp protection

The amount of suits with EMP protection has been reduced after talking with the maintainers, some key roles get EMP protection, and engineering has EMP proof cells they can use for fixing the SM or doing fusion / tesla / singulo.

**Comment by nightred** [comment]
I consider this functional and working as intended in all my tests.

File: code/modules/clothing/spacesuits/hardsuit.dm [hunk 2]
```diff
@@ -185,6 +187,13 @@
 	if(slot == ITEM_SLOT_OCLOTHING) //we only give the mob the ability to toggle the helmet if he's wearing the hardsuit.
 		return 1
 
+/// Burn the person inside the hard suit just a little, the suit got really hot for a moment
```
Comment (81Denton) [inline code/modules/clothing/spacesuits/hardsuit.dm]: This should have some sort of feedback message so people know why they're suddenly getting roasted

**Comment by cacogen** [comment]
testmerge this shit

File: code/modules/clothing/spacesuits/hardsuit.dm (hunk 2 above)
Comment (nightred) [reply code/modules/clothing/spacesuits/hardsuit.dm]: I added a notice to the wearer, and a chance to scream if you are burned.

**Comment by nightred** [comment]
Fixed up a merge conflict.

**Comment by optimumtact** [comment]
remove emp protection from CE, HOS SWAT CAP suits

**Pull request merged**